File: 06_go_back_n/GoBackN.java
// CN Lab Experiment 6: Go-Back-N Sliding Window Protocol
// Simple console simulation.

import java.util.Scanner;

public class GoBackN {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        System.out.print("Enter number of frames: ");
        int total = scanner.nextInt(); // Total frames.

        System.out.print("Enter window size: ");
        int window = scanner.nextInt(); // Number of frames per window.

        System.out.print("Enter lost frame number (0 for none): ");
        int lost = scanner.nextInt(); // Frame selected for simulated loss.

        int base = 1; // First unacknowledged frame.

        while (base <= total) { // Continue until every frame is delivered.
            int end = Math.min(base + window - 1, total); // Last frame in window.
            System.out.println("Sending frames " + base + " to " + end);

            if (lost >= base && lost <= end) { // Check for the simulated loss.
                System.out.println(
                    "Frame " + lost + " lost -> retransmit from frame " + lost
                );
                base = lost; // Go back to the lost frame.
                lost = 0;    // Assume retransmission succeeds.
            } else {
                System.out.println("ACK received for frames " + base + " to " + end);
                base = end + 1; // Move to the next window.
            }
        }

        System.out.println("All frames delivered successfully.");
        scanner.close();
    }
}

File: 06_go_back_n/go_back_n.py
# CN Lab Experiment 6: Go-Back-N Sliding Window Protocol
# This is a simple console simulation. A lost frame causes retransmission
# starting from that frame.

total = int(input("Enter number of frames: "))        # Total frames to send.
window = int(input("Enter window size: "))            # Number of frames in a window.
lost = int(input("Enter lost frame number (0 for none): "))  # Simulated loss.

base = 1                                               # First unacknowledged frame.

while base <= total:                                   # Continue until all frames are sent.
    end = min(base + window - 1, total)                # Find last frame in current window.
    print(f"Sending frames {base} to {end}")

    if lost >= base and lost <= end:                   # Check whether loss is in window.
        print(f"Frame {lost} lost -> Go-Back-N retransmits from frame {lost}.")
        base = lost                                    # Retransmit from the lost frame.
        lost = 0                                       # Simulate successful retransmission.
    else:
        print(f"ACK received for frames {base} to {end}.")
        base = end + 1                                 # Slide window forward.

print("All frames delivered successfully.")
